### ISSUE TITLE
prevent recording duplicate tv show episodes when the EPG provides the same episodes with different program IDs on different days

### ISSUE THREAD

**Issue by halucigenia** [MythTV/mythtv#332]
When the EPG provides the same episodes of a TV show with different program IDs shown on different times or days there is no way to prevent the same episodes from re-recording.

It would be useful if you could disable Program ID as the default criteria for finding duplicates and instead rely on the existing Duplicate Check methods only in these instances.

**Comment by stuarta** [comment]
It's interesting because program id's and series id's are supposed to be consistent for the same program, however unfortunately this is not the case. The EIT data transmitted by the broadcasters is typically the worst offender for doing this. So this request does make sense.

**Comment by kmdewaal** [comment]
This request does also make sense because the choices presented to the user should correspond with what the system can and will do. If the selection is made to do duplicate matching on title then that is what the system should do. So I think there should be another option added to select duplicate matching on program id.

On my system I have recordings with and without program id's and consequently duplicate matching fails. What happens is that the program id's are always available but only short before the recording starts in the Now & Next EIT tables. The next program is sometimes very short, e.g. a weather update or so. Combining that with a large pre-run this causes the recording to start before the program id is available and thus duplicate matching fails here. After the recording has started the program id information, when it finally arrives, is not updated anymore so the duplicate matching of the next occurrence of this program also fails.

Another consideration is that with a multi-tuner system usually only one tuner is configured to do EIT. When this tuner is used for a recording it will still get the EIT data that is present on that multiplex but that does not necessarily contain the Now & Next information for all available channels. 

It has been mentioned that Schedules Direct does have reliable program id information and that this can be a good argument to take a subscription. This may be true, but SD is not available in many places including where I live.

Finally, MythTV should work as good as possible with the broadcast signal. It cannot be that a feature such as duplicate detection is suddenly dropped to be replaced by a similar feature that is only available for money.

**Comment by gigem** [comment]
The duplicate checking is complicated enough already.  What we have now is the result of many iterations from a long, long time ago.  I don't see any need to try to "fix" it yet again.  If your program IDs are that unrelible, we can add options to not populate them in the EPG for particular sources.  We can also update the text in the schedule editor to clarify that the configurable duplicate checking is only done when program IDs aren't available.

**Comment by kmdewaal** [comment]
It is worthwhile to check if the program ID values as received from the broadcaster, when present, are correct. For my cable TV signal the values are always correct but the values are not always picked-up by MythTV. There are a number of reasons for this.
In my cable TV signal the program ID values are only present in the Now&Next (also called Present and Following) programs, which is the program now being broadcast and the next scheduled program. The programs after that do not yet have the program ID values. This causes problems in various ways, as described below.

When there are two very short programs followed by a third program that I want to record, and the third program is configured to start recording some time before the actual program start (the pre-roll) the recording will start before the program ID is available. If this is a duplicate the scheduler did not know it because the program ID was not yet available when the decision was made to start the recording.
In this case, even when the program ID becomes available after the recording has been started, the program ID is not stored in the database because the EIT data is ignored once the program is started. This means that the duplicate check for the next broadcast of the same program will also fail even if the program ID is then available.
This last bit looks like something that can be fixed in the EIT processing; it is feasible to update the program ID after the recording has started.
This then would also allow automatic duplicate removal after the recording has been completed, with some post-processing job.
It is even conceivable that a recording-in-progress is deleted immediately after it has been detected that it is a duplicate.

A second possible reason for missing program ID values is the current recommended practice of using only a single tuner per video source for EIT capturing. For my cable TV signal all transports have the complete EIT for all programs but that is not necessarily so. If the tuner is busy recording a program on a stream that has only limited or no EIT data then the other programs will not be updated while that recording is going on.
This can be fixed by enabling EIT for all tuners. Note that there is an issue with EIT on HDHomeRun tuners; there EIT only appears to work on the first tuner, as described in ticket https://code.mythtv.org/trac/ticket/13405. There has never been priority to fix this because using only one tuner is currently the recommended practice but this might be a good reason to fix this.

The third reason for missing program ID values is a bug in the EIT processing in conjunction with the multirec feature (recording more than one program from the same transport stream using only one tuner).
The following sequence illustrates the issue:
- The first recording starts. This also starts passive EIT processing (capturing EIT data from a transport that we are tuned to anyway).
- The second recording starts. There is no passive EIT processing because the EIT signal is already processed in the context of the first recording. This makes perfect sense because the EIT is not part of the program but is present only once in the stream. Processing exactly the same data at the same moment twice does not make sense.
- The first recording finishes. Stopping the recording also stops the passive EIT processing because, as mentioned before, the EIT processing is done in the context of a program recording. It is not possible to start active EIT processing (selecting a new channel and look at the EIT) because the tuner is still in use for the second program. The second program does not do EIT processing so as long as the second program is being recorded there will be no EIT information captured.
Enabling EIT capture on all tuners does somewhat alleviate this problem but will never completely solve this.

Disabling multirec can be an option if you have enough tuners anyway. On my HDHomeRun I only receive programs from one multiplex and these are now all recorded on one tuner, leaving the other 3 tuners idle, so this can be an option.
As this can be considered a bug the idea is to fix the EIT processing so that it does always do passive EIT when there are one or more recordings being made on a tuner.

For those cases where the EIT program ID values are really unreliable I do like the suggestion made by @gigem  in the previous post to completely ignore the program ID values for a particular video source. This would then be a mythtv-setup option.
However, when the program ID values are correct it is a much better method of duplicate detection that the title/description method so I think the priority should be to fix the EIT processing.

**Comment by cooldreamz** [comment]
A well known broadcaster is currently running "Tales of the Unexpected" and airing 6 episodes a day with each unique episode repeated three times. Unfortunately the programID is incremented each time any episode is broadcast. This is resulting in two duplicates for every episode. There are 112 unique episodes of "Tales of the Unexpected" so this will ultimately result in 224 duplicates.

Fortunately most broadcasters do a reasonable job of maintaining the programIDs in the EPG however the quality of the guide information does seem to have been getting progressively worse over the years.

The ability to selectively ignore programIDs in recording rules would solve this problem whilst still allowing 'most' rules to perform well.

An option to ignore programIDs for an entire source would be of very limited value and would probably cause more issues than it would solve. Setting it for an individual channel might be better (as some broadcasters are better than others) but the best approach would be to make this an option on a recording rule.
